### PR TITLE
gh-118789: Add `PyUnstable_Object_ClearWeakRefsNoCallbacks`

### PR DESCRIPTION
**Pull request by colesbury**
This exposes `_PyWeakref_ClearWeakRefsNoCallbacks` as an unstable C-API function to provide a thread-safe mechanism for clearing weakrefs without executing callbacks.

Some C-API extensions need to clear weakrefs without calling callbacks, such as after running finalizers like we do in subtype_dealloc. Previously they could use `_PyWeakref_ClearRef` on each weakref, but that's not thread-safe in the free-threaded build.

* Issue: gh-118789

----
📚 Documentation preview 📚: https://cpython-previews--118807.org.readthedocs.build/

### CODE REVIEW

**Comment by vstinner** [comment]
I suggest to rename the function to: `PyObject_ClearWeakRefs()`.

* I don't think that it deserves to belong to the unstable API, since it's stable for many years.
* It doesn't belong to PyWeakref API, since the argument is a generic object.
* I don't think that there is an use case to call the callbacks. If tomorrow, someone comes with such use case, we can add a new function. I don't think that adding a parameter for that would be useful right now.

File: Objects/weakrefobject.c [hunk 1]
```diff
@@ -1090,6 +1090,12 @@ _PyWeakref_ClearWeakRefsExceptCallbacks(PyObject *obj)
     UNLOCK_WEAKREFS(obj);
 }
 
+void
+PyUnstable_Weakref_ClearWeakRefsExceptCallbacks(PyObject *obj)
+{
```
Comment (vstinner) [inline Objects/weakrefobject.c]: I suggest to check if `_PyType_SUPPORTS_WEAKREFS(Py_TYPE(op))` and do nothing (return) if it's not the case.

**Comment by colesbury** [comment]
> I suggest to rename the function to: PyObject_ClearWeakRefs()

We already have `PyObject_ClearWeakRefs`: it clears the weakrefs *and* calls callbacks. It's the more commonly used API.

The use case for both functions is that types with user finalizers (i.e., `__del__` methods) need to perform the following sequence if they want to be correct:

* Clear weakrefs and call callbacks (i.e., `PyObject_ClearWeakRefs`)
* Call finalizer (i.e, `PyObject_CallFinalizerFromDealloc`)
* Clear weakrefs *again*, but don't call callbacks (i.e., `PyUnstable_Weakref_ClearWeakRefsExceptCallbacks`)

The third step avoids leaking memory in case the finalizer added new weakrefs. We do this internally in [`subtype_dealloc`](https://github.com/python/cpython/blob/35b5eaa176a5bae8a0cacb9c9f40ec948ecc4325/Objects/typeobject.c#L2250-L2258), but extensions can't always delegate to `subtype_dealloc`.

Previously, the third step was implemented as:

```
      PyWeakReference** list =
          (PyWeakReference**)PyObject_GET_WEAKREFS_LISTPTR(self);
      while (*list)
        _PyWeakref_ClearRef(*list);
    }
```

But that's not thread-safe without the GIL.

> It doesn't belong to PyWeakref API, since the argument is a generic object

I'll adjust the name and add the check you suggest.

File: Doc/c-api/weakref.rst [hunk 2]
```diff
@@ -96,3 +96,10 @@ as much as it can.
    This iterates through the weak references for *object* and calls callbacks
    for those references which have one. It returns when all callbacks have
    been attempted.
+
+
+.. c:function:: void PyUnstable_Object_ClearWeakRefsExceptCallbacks(PyObject *object)
+
+   Clears the weakrefs for *object* without calling the callbacks.
+
+   .. versionadded:: 3.13
```
Comment (vstinner) [inline Doc/c-api/weakref.rst:114]: IMO it's too late for 3.13, we're past beta1.

```suggestion
   .. versionadded:: 3.14
```

**Comment by vstinner** [comment]
Why do you want to put it in the PyUnstable API?

**Comment by colesbury** [comment]
>  Why do you want to put it in the PyUnstable API?

That's what I understood from @Yhg1s's comments on Discord (#c-api) as the appropriate name/stability for 3.13.

I'd certainly prefer something outside of the PyUnstable API as the end state for this (i.e., in 3.14).

**Comment by vstinner** [comment]
I suggest to only do the revert in 3.13, and add the function only in 3.14.

**Comment by colesbury** [comment]
If we only add the function in 3.14, then there will be no way for C-API extensions that need this functionality to be made thread-safe in 3.13.

**Comment by colesbury** [comment]
I'll open a https://github.com/capi-workgroup/decisions issue for 3.14

**Comment by colesbury** [comment]
@encukou, would you please review this?

File: Doc/c-api/weakref.rst [hunk 3]
```diff
@@ -96,3 +96,10 @@ as much as it can.
    This iterates through the weak references for *object* and calls callbacks
    for those references which have one. It returns when all callbacks have
    been attempted.
+
+
+.. c:function:: void PyUnstable_Object_ClearWeakRefsNoCallbacks(PyObject *object)
+
+   Clears the weakrefs for *object* without calling the callbacks.
```
Comment (vstinner) [inline Doc/c-api/weakref.rst:103]: Would you mind to document when and where this function is expected to be called? You described it here: https://github.com/capi-workgroup/decisions/issues/25#issue-2290152853 ("... Clear weakrefs again, but don't call callbacks ...").

I understood that it's mostly for deallocator functions, no?

I suggest to recommend using PyObject_ClearWeakRefs() in general.

**Comment by encukou** [comment]
Unstable API needs tests; see https://github.com/colesbury/cpython/pull/2. Otherwise, this looks good to me.


@Yhg1s, do you want this in 3.13? As mentioned in the issue, projects that reimplement `subtype_dealloc` need this to be thread-safe in free-threading builds.
The proper solution is (IMO) to give them API so they don't need to reimplement `subtype_dealloc`, but that's definitely not coming in 3.13.

File: Doc/c-api/weakref.rst (hunk 2 above)
Comment (colesbury) [reply Doc/c-api/weakref.rst:114]: @Yhg1s - is it okay to make this function public in 3.13? From our discussion on Discord in May, it sounded like you were okay with this before the RC, but I'd like to make sure.

File: Doc/c-api/weakref.rst (hunk 3 above)
Comment (colesbury) [reply Doc/c-api/weakref.rst:103]: I've updated the docs

**Review by vstinner — APPROVED** [review]
LGTM. For the version number (`.. versionadded::` in the doc), it depends if you backport the change to 3.13. If no, you should use 3.14.

File: Doc/c-api/weakref.rst (hunk 2 above)
Comment (Yhg1s) [reply Doc/c-api/weakref.rst:114]: Yes, considering the circumstances, this is reasonable to add in beta 3.

**Review by Yhg1s — APPROVED** [review]

**Comment by vstinner** [comment]
Congrats @colesbury :-) PyTorch should be happy with this new API if I understood correctly.

**Pull request merged**